src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -1445,4 +1519,122 @@ ObjectDataHandler::clear_ret ObjectDataHandler::clear(
     });
 }
 
+ObjectDataHandler::clone_ret ObjectDataHandler::clone_extents(
+  context_t ctx,
+  object_data_t &object_data,
+  lba_pin_list_t &pins,
+  laddr_t data_base)
+{
+  LOG_PREFIX(ObjectDataHandler::clone_extents);
+  TRACET(" object_data: {}~{}, data_base: {}",
+    ctx.t,
+    object_data.get_reserved_data_base(),
+    object_data.get_reserved_data_len(),
+    data_base);
+  return ctx.tm.dec_ref(
+    ctx.t,
+    object_data.get_reserved_data_base()
+  ).si_then(
+    [&pins, &object_data, ctx, data_base](auto) mutable {
+      return seastar::do_with(
+	(extent_len_t)0,
+	[&object_data, ctx, data_base, &pins](auto &last_pos) {
+	return trans_intr::do_for_each(
+	  pins,
+	  [&last_pos, &object_data, ctx, data_base](auto &pin) {
+	  auto offset = pin->get_key() - data_base;
+	  ceph_assert(offset == last_pos);
+	  auto fut = TransactionManager::alloc_extent_iertr
+	    ::make_ready_future<LBAMappingRef>();
+	  auto addr = object_data.get_reserved_data_base() + offset;
+	  if (pin->get_val().is_zero()) {
+	    fut = ctx.tm.reserve_region(ctx.t, addr, pin->get_length());
+	  } else {
+	    fut = ctx.tm.clone_extent(ctx.t, addr, *pin);
+	  }
+	  return fut.si_then(
+	    [&pin, &last_pos, offset](auto) {
+	    last_pos = offset + pin->get_length();
+	    return seastar::now();
+	  }).handle_error_interruptible(
+	    crimson::ct_error::input_output_error::pass_further(),
+	    crimson::ct_error::assert_all("not possible")
+	  );
+	}).si_then([&last_pos, &object_data, ctx] {
+	  if (last_pos != object_data.get_reserved_data_len()) {
+	    return ctx.tm.reserve_region(
+	      ctx.t,
+	      object_data.get_reserved_data_base() + last_pos,
+	      object_data.get_reserved_data_len() - last_pos
+	    ).si_then([](auto) {
+	      return seastar::now();
+	    });
+	  }
+	  return TransactionManager::reserve_extent_iertr::now();
+	});
+      });
+    },
+    ObjectDataHandler::write_iertr::pass_further{},
+    crimson::ct_error::assert_all{
+      "object_data_handler::clone invalid error"
+    }
+  );
+}
+
+ObjectDataHandler::clone_ret ObjectDataHandler::clone(
+  context_t ctx)
+{
+  // the whole clone procedure can be seperated into the following steps:
+  // 	1. let clone onode(d_object_data) take the head onode's
+  // 	   object data base;
+  // 	2. reserve a new region in lba tree for the head onode;
+  // 	3. clone all extents of the clone onode, see transaction_manager.h
+  // 	   for the details of clone_extent;
+  // 	4. reserve the space between the head onode's size and its reservation
+  // 	   length.
+  return with_objects_data(
+    ctx,
+    [ctx, this](auto &object_data, auto &d_object_data) {
+    ceph_assert(d_object_data.is_null());
+    return prepare_data_reservation(
+      ctx,
+      d_object_data,
+      object_data.get_reserved_data_len()
+    ).si_then([&object_data, &d_object_data, ctx, this] {
+      auto base = object_data.get_reserved_data_base();
+      auto len = object_data.get_reserved_data_len();
```

Comment:
`object_data` is not null after `prepare_data_reservation`.

Reply:
L1601 above is preparing `d_object_data`, or is it else where?

Reply:
Ah, you're right. I'll add an assertion here.

Reply:
>  I'll add an assertion here.

Should there be another `prepare_data_reservation` to guarantee that the input `object_data` is initialized?

Reply:
I think OP_CLONE can only happen on a object that has data, so we shouldn't reserve lba space here.

Reply:
I don't see how OP_CLONE followed by OP_CREATE/TOUCH is impossible from store perspective. IIRC the newly created object does have data but they are all 0s, and it is lazy initialized as a null object.

Or am I missing anything?

Reply:
> I don't see how OP_CLONE followed by OP_CREATE/TOUCH is impossible from store perspective.

I think you mean OP_CREATE/TOUCH followed by OP_CLONE, right?

Anyway, even in this case, I don't think we should reserve a piece of lba space just because we are doing OP_CLONE. Probably we should check the "nullness" of `object_data` at the beginning of `ObjectDataHandler::clone()` and exit if it's null.